### ISSUE TITLE
Slow cUserFunctionPointer highlighting

### ISSUE THREAD

**Issue by felixendres** [bfrg/vim-cpp-modern#36]
Hello,
I have experienced long slowdowns, when editing C++ code, which made typing very cumbersome.
It is probably a combination of things, but the thing that helped in the end was removing that line: 

https://github.com/bfrg/vim-cpp-modern/blob/1e6d2b52e34f222f79ecf523643a794f9079d7bd/after/syntax/c.vim#L21

I had profiled the syntax highlighting using
- :syntime on
- Ctrl-L for redraw
- :syntime off
- :syntime report

The result was a 5 second total for cUserFunctionPointer alone.

Not sure how to better solve it, but since I couldn't find anything about this issue on the web, I thought it may be interesting for other people, and maybe someone here can give me a hint. 

Some information about my setup:
- gvim 9.0 with patches 1-749
- 2 C++ files open in a vertical split in a tab
- Files are 150 and 195 lines long
- Windows were 110 lines tall

**Comment by bfrg** [comment]
Thanks for reporting this. Indeed it's a quite complex regex pattern that includes a look-behind and a `\_s` (i.e. whitespace with newlines) atomic group. Personally, I haven't experienced any slowdown. So I'm curious, what kind of CPU do you have and which OS?

Could you try it by replacing the `\_s` with a `\s` in the syntax file, i.e.
```vim
syn match cUserFunctionPointer "\%((\s*\*\s*\)\@<=\h\w*\ze\s*)\s*(.*)"
```

If that doesn't improve anything, it must be the (variable-width) look-behind. We could add a limit here, for example, to look back only 6 bytes:
```vim
syn match cUserFunctionPointer "\%((\s*\*\s*\)\@6<=\h\w*\ze\s*)\_s\{-}(.*)"
```

This is actually even recommended in Vim's documentation under [`:h /\@<=`](https://github.com/vim/vim/blob/03acd4761be1c2766d3ec17534ea63cdf8dd565d/runtime/doc/pattern.txt#L736-L739).

**Comment by felixendres** [comment]
Thank you, I've tried both suggestions. They both seem to reduce the timings from syntime, but the measurements are very erratic, so it's hard to say exactly.

The weird thing is, that it mostly seems to slow down, when I edit in a C++ comment section and autocompletion (using vim-lsp) is active. I noticed that with vim-lsp enabled, the counts (2nd column) in the output of :syntime report is an order of magnitude higher than what it normally is. E.g. 44k instead of 3k. So I guess that aggravates the situation in my case and I'll have to figure out, why the counts are so high. Do you have any idea about this?

I have a fairly performant system (Ubuntu 20.04 on an 11th Gen Intel Core i7-11850H @ 2.50GHz), so that shouldn't be the problem.

Anyway, the delay is noticably quicker now, so thank you again.

**Comment by bfrg** [comment]
> The weird thing is, that it mostly seems to slow down, when I edit in a C++ comment section and autocompletion (using vim-lsp) is active. I noticed that with vim-lsp enabled, the counts (2nd column) in the output of :syntime report is an order of magnitude higher than what it normally is. E.g. 44k instead of 3k. So I guess that aggravates the situation in my case and I'll have to figure out, why the counts are so high. Do you have any idea about this?

Do you observe any slowdowns when vim-lsp is disabled? AFAIR vim-lsp sends diffs to the LSP server on every text change (I think, using timers). Determining the diffs can be computationally costly since it's done on every character change. At the same time, if you're using its auto-completion, the list of completion candidates needs to be determined as well on every text change. I think, they're even using Vim's matchfuzzy function, which is also computationally heavy. And since Vim doesn't support multi-threading, everything needs to be done in the main event loop, together with its internal syntax engine.

Have you tried another LSP client? For example: https://github.com/yegappan/lsp. It uses vim9script which is significantly faster than the old vimscript. There's also a vim9script auto-completion plugin: https://github.com/girishji/vimcomplete that supports yegappan/lsp.

**Comment by felixendres** [comment]
> Have you tried another LSP client?

I have to correct myself, I am actually using yegappan/lsp, as you suggested. I used vim-lsp last year, but had some issues with it.

Also, I have to correct myself with the syntime count and totals, I forgot to do `syntime clear`  in between.

> Do you observe any slowdowns when vim-lsp is disabled? 

~~Yes, but much less. With lsp's completion popup, the lag is a few seconds. Without lsp, it is noticable, but less than half a second (had to judge numbers below that). However, it's annoying enough, that I will keep the cUserFunctionPointer regex disabled, even with the suggested change. If I do `syntax off` things are smooth in both cases (lsp on/off).~~

Things changed after a reboot. I do not seem to get consistent results, not sure why. Now there's no slowdown without lsp anymore, even with the old cUserFunctionPointer.

I still get the long delays with lsp completion enabled, but not when syntax is off. So there is some interplay between them, which I don't understand. But I guess the cUserFunctionPointer regex is not the culprit.

Sorry for the mess, thanks for your time.  I'll try vimcomplete.

**Comment by felixendres** [comment]
It's been a while, but I have found the main culprit: `foldmethod=syntax`

I checked the effect of options and plugins by looking at the `syntime report`. Concretely, I compared the number of calls of the  syntax item regex (the second column of `syntime report`) of the slowest (i.e. top) entry in the report.

I measured when inserting "12345" into the middle of a C++ comment.

- Without any vim profile (vim -u NONE) the number of calls to the top regex was 10.
- Adding `set conceallevel=2` made it jump from 10 to 18
- The plugin "markonm/traces.vim" made the value go from 10 to ~95
- Setting the foldmethod to syntax makes it go from 10 to 2300-2500. Regardless of the foldlevel.
